timelock/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+                timestamp,
+            )
+                .into_val(&env),
+        );
+        // Transfer token to this contract address.
+        transfer_from(&env, &token, &from_id, &get_contract_id(&env), &amount);
+        // Store all the necessary balance to allow one of the claimants to claim it.
+        env.contract_data().set(
+            DataKey::Balance,
+            ClaimableBalance {
+                token,
+                amount,
+                time_bound: TimeBound {
+                    kind: time_bound_kind,
+                    timestamp,
+                },
+                claimants,
+            },
+        );
+    }
+
+    pub fn claim(env: Env, claimant: Signature) {
+        if !is_initialized(&env) {
+            panic!("contract hasn't been initialized");
+        }
+        let claimable_balance: ClaimableBalance =
+            env.contract_data().get_unchecked(DataKey::Balance).unwrap();
+
+        if !check_time_bound(&env, &claimable_balance.time_bound) {
+            panic!("time predicate is not fulfilled");
+        }
+
+        let claimant_id = claimant.get_identifier(&env);
+        let claimants = &claimable_balance.claimants;
+        // Check if claimaint is one of the allowed claimants.
+        // `contains` is not implemented for Soroban vec (yet).
+        if !claimants.iter().any(|c| c.unwrap() == claimant_id) {
+            panic!("claimant is not allowed to claim this balance");
+        }
+        // Authenticate claimant with nonce of zero, so that this may be
+        // successfully called just once.
+        // For simplicity, depositor can't be the claimant.
+        check_auth(
+            &env,
+            &NonceForSignature(claimant),
+            BigInt::zero(&env),
+            Symbol::from_str("claim"),
+            (&claimant_id,).into_val(&env),
+        );
+        // Transfer the stored amount of token to claimant after passing
+        // all the checks.
+        transfer_to(
+            &env,
+            &claimable_balance.token,
+            &claimant_id,
+            &claimable_balance.amount,
+        );
+        // Cleanup unnecessary balance entry.
+        env.contract_data().remove(DataKey::Balance);
+    }
+}
+
+fn is_initialized(env: &Env) -> bool {
+    env.contract_data().has(DataKey::Balance)
+}
+
+fn get_contract_id(e: &Env) -> Identifier {
+    Identifier::Contract(e.get_current_contract().into())
+}
+
+fn transfer_from(
+    e: &Env,
+    token_id: &BytesN<32>,
+    from: &Identifier,
+    to: &Identifier,
+    amount: &BigInt,
+) {
+    let client = TokenClient::new(&e, token_id);
+    let nonce = client.nonce(&get_contract_id(&e));
+    client.xfer_from(&Signature::Contract, &nonce, &from, &to, &amount);
+}
+
+fn transfer_to(e: &Env, token_id: &BytesN<32>, to: &Identifier, amount: &BigInt) {
+    let client = TokenClient::new(&e, token_id);
+    let nonce = client.nonce(&get_contract_id(&e));
+    client.xfer(&Signature::Contract, &nonce, to, amount);
```

Comment:
Same as comment above. You can pass in 0.

Reply:
We use contract nonce, hence this isn't 0, is it? Strictly speaking it should be 1 for this call when things go as intended.